### SRA Fetcher

This script was built for bulk downloading of SRA files containing Illumina Short Reads from a list of NGS projects.
The goal of this job was to get a huge number of sequences from Mycobacterium tuberculosis sequencing projects around the world in order to compare them, 
their variability, get a consensus sequence, detect possible super-mutable regions and associate SNPs to strains and lineages across the globe.

For that, I first created a bash script that uses [eutils](https://www.ncbi.nlm.nih.gov/books/NBK25500/) from NCBI to batch download the files we were interested in, 
and then it uses [autosnippy](https://github.com/MG-IiSGM/autosnippy) an all-in-one tool from our lab that combines read mapping, variant calling, species 
and lineage detection, annotation and philogenetic tree construction.

After that, a script in Python parses the data generated by autosnippy to:
1. Make a SNP database and count the number of new SNPs per sample, that apparently follow a [Heaps' Law](https://en.wikipedia.org/wiki/Heaps%27_law), (see Fig. 1).
2. Spot the ultra-variable sequences, by plotting an histogram with the number of SNPs each 100 pb on the M.tuberculosis chromosome (see Fig. 2).
3. Builds an antibiotic marker-to-lineage dictionary.

![heaps](https://user-images.githubusercontent.com/77884314/158214595-87dc9c32-3fd2-476b-a55d-03ab0d49ea2a.png)
*Figure 1*

![hyperv](https://user-images.githubusercontent.com/77884314/158216088-b7d24db2-8297-4e0e-bb30-e56d864e2de6.png)
*Figure 2*
